packages/@react-spectrum/combobox/test/ComboBox.test.js

```diff
@@ -304,83 +307,66 @@ describe('ComboBox', function () {
     });
     await user.keyboard('One');
-    act(() => {
-      jest.runAllTimers();
-    });
-
-    expect(queryByRole('listbox')).toBeNull();
+    expect(getListbox()).toBeFalsy();
     expect(onOpenChange).not.toHaveBeenCalled();
     expect(onFocus).not.toHaveBeenCalled();
 
-    await user.keyboard('{ArrowDown}');
-    act(() => {
-      jest.runAllTimers();
-    });
+    setInteractionType('keyboard');
+    await open();
 
-    expect(queryByRole('listbox')).toBeNull();
+    expect(getListbox()).toBeFalsy();
     expect(onOpenChange).not.toHaveBeenCalled();
 
-    let button = getByRole('button');
-    await user.click(button);
-    act(() => {
-      jest.runAllTimers();
-    });
-
-    expect(queryByRole('listbox')).toBeNull();
+    setInteractionType('mouse');
+    await open();
+    expect(getListbox()).toBeFalsy();
     expect(onOpenChange).not.toHaveBeenCalled();
     expect(onInputChange).not.toHaveBeenCalled();
   });
 
   it('can be readonly', async function () {
-    let {getByRole, queryByRole} = renderComboBox({isReadOnly: true, defaultInputValue: 'Blargh'});
+    let tree = renderComboBox({isReadOnly: true, defaultInputValue: 'Blargh'});
+    let {setInteractionType, setElement, getCombobox, getListbox, open} = testUtilUser.createTester('ComboBoxTester');
+    setElement(tree.container);
 
-    let combobox = getByRole('combobox');
     act(() => {
-      combobox.focus();
+      getCombobox().focus();
     });
     await user.keyboard('One');
-    act(() => {
-      jest.runAllTimers();
-    });
-
-    expect(queryByRole('listbox')).toBeNull();
-    expect(combobox.value).toBe('Blargh');
+    expect(getListbox()).toBeFalsy();
+    expect(getCombobox().value).toBe('Blargh');
     expect(onOpenChange).not.toHaveBeenCalled();
     expect(onFocus).toHaveBeenCalled();
     expect(onInputChange).not.toHaveBeenCalled();
 
-    await user.keyboard('{ArrowDown}');
-    act(() => {
-      jest.runAllTimers();
-    });
+    setInteractionType('keyboard');
+    await open();
 
-    expect(queryByRole('listbox')).toBeNull();
+    expect(getListbox()).toBeFalsy();
     expect(onOpenChange).not.toHaveBeenCalled();
     expect(onInputChange).not.toHaveBeenCalled();
 
-    let button = getByRole('button');
-    // user event click on the button actually makes focus leave the combobox
-    // so onInputChange will fire
-    await user.click(button);
-    act(() => {
-      jest.runAllTimers();
-    });
+    setInteractionType('mouse');
+    await open();
 
-    expect(queryByRole('listbox')).toBeNull();
+    expect(getListbox()).toBeFalsy();
     expect(onOpenChange).not.toHaveBeenCalled();
   });
 
   it('features default behavior of completionMode suggest and menuTrigger input', async function () {
-    let {getByRole} = renderComboBox();
+    let tree = renderComboBox();
+    let {setElement, getCombobox, getListbox, getOptions, getFocusedOption} = testUtilUser.createTester('ComboBoxTester');
+    setElement(tree.container);
```

Comment:
I expect that users would typically use a `data-testid` but there could be many different identifiers that someone could be using (label text like you mentioned , text content, an id, etc) so I didn't want to be restrictive. We could certainly add some additional options but I figured users would at a minimum be able to communicate the element they want to test.

Reply:
instead of setElement, i think `root` makes more sense since that's where we'll start searching from. it's really incidental that if the actual element we want is passed along as the root.